### ISSUE TITLE
Improve exception handling

### ISSUE THREAD

**Issue by alexis-via** [invoice-x/invoice2data#190]
I'm using invoice2data as a python lib in the Odoo module account_invoice_import_invoice2data, cf https://github.com/OCA/edi/blob/10.0/account_invoice_import_invoice2data/wizard/account_invoice_import.py#L54

When the extraction fails because 

1) some fields didn't match
2) or because no template was found

then the extract_data method of the lib just returns nothing and the reason of the failure is written in the logs.

In order to display an accurate error message to the user of Odoo, writing the reason of the failure in the logs is not enough, I think we need proper exceptions with nice error messages in exceptions so that I can catch these exceptions and display a good error message.

**Comment by alexis-via** [comment]
I don't have a lot of experience with python exceptions, so I may not be the best person to advise on how we should design the exceptions for invoice2data. @hbrunn could you give us some advice on this ? As you have a great Python+Odoo experience and you already contributed to this project, your input would be very appreciated.

**Comment by hbrunn** [comment]
I'd start by declaring some base class deriving from ValueError (or some other standard exception), and simply add a constructor that allows you to pass whatever data you need to pass. Exceptions are just class instances, so you can add whatever you like. cf https://docs.python.org/2/library/exceptions.html#exceptions.UnicodeError for an example in the standard lib

**Comment by hbrunn** [comment]
PS: Don't forget to override `__str__` to print something sensible using those values

**Comment by m3nu** [comment]
Anyone working on this? If yes, please assign yourself to this issue, so it's "reserved".

**Comment by bosd** [comment]
I agree with @alexis-via on this one. Some additional error message is very welcomed.
Even a function that when for instace one input parser fails (e.g. pdftotext). Odoo provides the user with the option to use another parser (tesseract).

@rmilecki Any input on this? Are you using invoice2data as a lib?

**Comment by rmilecki** [comment]
I use it as a lib but whenever something goes wrong, I have to use CLI to debug. So I'm all for adding some better errors handling.

**Comment by bosd** [comment]
Status on `next`: extraction now uses exceptions **internally** — `_check_required_fields` raises `ValueError('Unable to parse required field(s): ...')` and the input-backend cascade catches it to fall through to the next backend / OCR. But the **public `extract_data` still returns `{}`** on total failure, so a library caller (e.g. the Odoo module) still can't catch a typed reason — exactly the problem you raised.

Proposed for the 1.0.0 breaking window:
- Add an exception hierarchy: `InvoiceProcessingError(Exception)` with `NoTemplateFound` and `RequiredFieldsMissing` (the latter carrying the missing field names) — subclassing `ValueError` so the cascade's existing `except ValueError` keeps working.
- Surface it at the boundary either by (a) making `extract_data` raise on failure, or (b) an opt-in `raise_on_error=True` parameter so existing `{}`-checking callers don't break.

I lean toward (b) for safety. @bosd @alexis-via — which contract do you want and I'll implement it.

**Comment by bosd** [comment]
Implemented option (b) on `next` (941e297), ships in 1.0.0.

`extract_data(..., raise_on_error=True)` now raises a typed exception instead of returning `{}`:

```python
from invoice2data import extract_data, NoTemplateFoundError, RequiredFieldsMissingError

try:
    data = extract_data(path, raise_on_error=True)
except RequiredFieldsMissingError as exc:
    show_user(f"Matched a template but couldn't read: {', '.join(exc.fields)}")
except NoTemplateFoundError:
    show_user("No template matched this document.")
```

- New `invoice2data.exceptions`: `InvoiceProcessingError` (base), `NoTemplateFoundError`, `RequiredFieldsMissingError` (carries `.fields` + `.template_name`). All re-exported from `invoice2data`.
- `RequiredFieldsMissingError` subclasses `ValueError`, so the input-backend cascade's existing `except ValueError` retry is unaffected.
- **Non-breaking**: the default is still `raise_on_error=False` → returns `{}`, so existing callers (incl. the Odoo module) are unchanged until they opt in.
- (Names carry the `Error` suffix to satisfy the project's `pep8-naming` lint — slightly different from the draft in my earlier comment.)

Documented in the API reference, usage, and the 1.0 migration guide. @alexis-via this should give you catchable, message-bearing failures for the Odoo import wizard.
